File: 02-arrays/medium/04-kadane.cpp
#include <bits/stdc++.h>
using namespace std;
// Kadane's Algorithm : Maximum Subarray Sum in an Array
// Problem Statement : Given an integer array arr, find the contiguous
// subarray(containing at least one number) which has the largest sum and
// returns its sum and prints the subarray.
inline long long maxSubarraySum(vector<int> &arr) {
    int n = arr.size();
    long long maxi = LONG_MIN;
    long long sum = 0;
    int start = 0;
    int maxStart = -1;
    int maxEnd = -1;
    for (int i = 0; i < n; i++) {
        if (sum == 0)
            start = i;
        sum += arr[i];
        if (sum > maxi) {
            maxStart = start;
            maxEnd = i;
            maxi = sum;
        }
        if (sum < 0)
            sum = 0;
    }
    for (int i = maxStart; i <= maxEnd; i++)
        cout << arr[i] << ",";
    cout << endl;
    return maxi;
}

File: 02-arrays/medium/04-kadane.go
package main

import (
	"fmt"
	"math"
)

func SubarraySum(arr []int) (int, []int) {
	sum := 0
	maxi := math.MinInt
	start := 0
	end := 0
	for i := 0; i < len(arr); i++ {
		if sum == 0 {
			start = i
		}
		sum += arr[i]
		if sum > maxi {
			maxi = sum
			end = i
		}
		if sum < 0 {
			sum = 0
		}
	}
	return maxi, arr[start : end+1]
}

func main() {
	arr := []int{6, -2, -3, 2, -1, 4, -1, -2, 1, 5, -3}
	max, maxarr := SubarraySum(arr)
	fmt.Println(max)
	fmt.Println(maxarr)

}
